supplier_app/views.py

```diff
@@ -199,58 +215,73 @@ def get_context_data(self, **kwargs):
         return context
 
 
-class EditTaxpayerView(UserLoginPermissionRequiredMixin, UpdateView):
+class EditTaxpayerView(UserLoginPermissionRequiredMixin, HasTaxPayerPermissionMixin, UpdateView):
     template_name = 'AP_app/edit-taxpayer-information.html'
     model = TaxPayerArgentina
     form_class = TaxPayerEditForm
     pk_url_kwarg = "taxpayer_id"
     permission_required = (
-        'users_app.can_edit_taxpayer',
+        CAN_EDIT_TAXPAYER_PERM,
     )
 
     def get_context_data(self, **kwargs):
         context = super().get_context_data(**kwargs)
         context['taxpayer_id'] = self.kwargs['taxpayer_id']
         return context
 
+    def handle_no_permission(self):
+        return HttpResponseRedirect(Http404)
+
     def get_success_url(self, **kwargs):
         taxpayer_id = self.kwargs['taxpayer_id']
         return reverse('supplier-details', kwargs={'taxpayer_id': taxpayer_id})
 
 
-class EditAddressView(UserLoginPermissionRequiredMixin, UpdateView):
+class EditAddressView(UserLoginPermissionRequiredMixin, HasTaxPayerPermissionMixin, UpdateView):
     template_name = 'AP_app/edit-address-information.html'
     model = Address
     form_class = AddressCreateForm
     pk_url_kwarg = "address_id"
     permission_required = (
-        'users_app.can_edit_taxpayer',
+        CAN_EDIT_TAXPAYER_PERM,
+        CAN_EDIT_TAXPAYER_ADDRESS_PERM,
     )
 
     def get_context_data(self, **kwargs):
         context = super().get_context_data(**kwargs)
-        context['taxpayer_id'] = Address.objects.get(pk=self.kwargs['address_id']).taxpayer.id
+        taxpayer = get_object_or_404(TaxPayer, id=self.kwargs['taxpayer_id'])
+        context['taxpayer_id'] = taxpayer.id
+        get_object_or_404(Address, pk=self.kwargs['address_id'], taxpayer=taxpayer)
         return context
 
+    def handle_no_permission(self):
+        return HttpResponseRedirect(Http404)
+
     def get_success_url(self, **kwargs):
         taxpayer_id = Address.objects.get(pk=self.kwargs['address_id']).taxpayer.id
         return reverse('supplier-details', kwargs={'taxpayer_id': taxpayer_id})
 
 
-class EditBankAccountView(UserLoginPermissionRequiredMixin, UpdateView):
+class EditBankAccountView(UserLoginPermissionRequiredMixin, HasTaxPayerPermissionMixin, UpdateView):
```

Comment:
Seems little weird to have `HasTaxPayerPermissionMixin`, I think if we are talking about a Mixin, it should be only `TaxPayerPermissionMixin` or something similar. Please contact me if you have doubts

Reply:
Done